common/src/main/java/io/netty5/util/concurrent/DefaultTaskScheduler.java

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright 2023 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty5.util.concurrent;
+
+import io.netty5.util.concurrent.AbstractScheduledEventExecutor.RunnableScheduledFutureNode;
+import io.netty5.util.internal.DefaultPriorityQueue;
+import io.netty5.util.internal.PriorityQueue;
+
+import java.util.Comparator;
+import java.util.concurrent.Callable;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Default task scheduler for {@link AbstractScheduledEventExecutor} that want to support task scheduling.
+ * Uses {@link DefaultPriorityQueue} to schdeule tasks.
+ */
+public final class DefaultTaskScheduler extends AbstractTaskScheduler {
+
+    private static final Comparator<RunnableScheduledFutureNode<?>> SCHEDULED_FUTURE_TASK_COMPARATOR =
+            Comparable::compareTo;
+    private static final RunnableScheduledFutureNode<?>[]
+            EMPTY_RUNNABLE_SCHEDULED_FUTURE_NODES = new RunnableScheduledFutureNode<?>[0];
+
+    private final AbstractScheduledEventExecutor executor;
+    private final PriorityQueue<RunnableScheduledFutureNode<?>> scheduledTaskQueue;
+
+    public DefaultTaskScheduler(AbstractScheduledEventExecutor executor) {
+        this.executor = requireNonNull(executor, "executor");
+        scheduledTaskQueue = new DefaultPriorityQueue<>(
+                SCHEDULED_FUTURE_TASK_COMPARATOR,
+                // Use same initial capacity as java.util.PriorityQueue
+                11);
+    }
+
+    @Override
+    protected <V> RunnableScheduledFuture<V> newScheduledTaskFor(Callable<V> callable, long deadlineNanos,
+                                                                 long period) {
+        return new RunnableScheduledFutureAdapter<>(executor, executor.newPromise(), callable, deadlineNanos, period);
```

Comment:
**Performance Issue**: The RunnableScheduledFutureAdapter instantiation within DefaultTaskScheduler's newScheduledTaskFor method doesn't consider the reuse of instances, which might lead to unnecessary object creation and increased garbage collection overhead. <br> **Fix**: Implement a mechanism to reuse RunnableScheduledFutureAdapter instances, such as an object pool. This approach will reduce the number of created objects and mitigate the garbage collection impact, especially under high load. <br> **Code Suggestion**: 
 ```
 private final ObjectPool<RunnableScheduledFutureAdapter<?>> runnableScheduledFutureAdapterPool = new DefaultObjectPool<>(new ObjectPool.ObjectCreator<RunnableScheduledFutureAdapter<?>>() {
     @Override
     public RunnableScheduledFutureAdapter<?> create() {
         return new RunnableScheduledFutureAdapter<>(DefaultTaskScheduler.this, newPromise(), callable, deadlineNanos, periodNanos);
     }
 });

 @Override
 protected <V> RunnableScheduledFuture<V> newScheduledTaskFor(Callable<V> callable, long deadlineNanos, long periodNanos) {
     return (RunnableScheduledFuture<V>) runnableScheduledFutureAdapterPool.get().init(callable, deadlineNanos, periodNanos);
 }
 ```